Sources/Basics/Observability.swift

```diff
@@ -0,0 +1,293 @@
+            return DiagnosticsEmitter(context: context).emit(error)
+        }
+        self.emit(.error(error, data: data))
+    }
+
+    public func emit(warning message: String, data: CustomStringConvertible? = .none) {
+        self.emit(.warning(message, data: data))
+    }
+
+    public func emit(info message: String, data: CustomStringConvertible? = .none) {
+        self.emit(.info(message, data: data))
+    }
+
+    public func trap<T>(_ closure: () throws -> T) -> T? {
+        do  {
+            return try closure()
+        } catch {
+            self.emit(error)
+            return nil
+        }
+    }
+}
+
+public struct DiagnosticMessage: Equatable {
+    let severity: Severity
+    let message: String
+    // 👀 TODO: not sure if this is used very much, but it we need it this could/should be changed to more structured metadata model
+    let data: CustomStringConvertible?
+
+    public static func error(_ message: String, data: CustomStringConvertible? = .none) -> Self {
+        Self(severity: .error, message: message, data: data)
+    }
+
+    public static func error(_ message: CustomStringConvertible, data: CustomStringConvertible? = .none) -> Self {
+        Self(severity: .error, message: message.description, data: data)
+    }
+
+    public static func error(_ error: Error, data: CustomStringConvertible? = .none) -> Self {
+        let message: String
+        // FIXME: this brings in the TSC API still
+        // FIXME: string interpolation seems brittle
+        if let diagnosticData = error as? DiagnosticData {
+            message = "\(diagnosticData)"
+        } else if let convertible = error as? DiagnosticDataConvertible {
+            message = "\(convertible.diagnosticData)"
+        } else {
+            message = "\(error)"
+        }
+        return Self(severity: .error, message: message, data: data)
+    }
+
+    public static func warning(_ message: String, data: CustomStringConvertible? = .none) -> Self {
+        Self(severity: .warning, message: message, data: data)
+    }
+
+    public static func warning(_ message: CustomStringConvertible, data: CustomStringConvertible? = .none) -> Self {
+        Self(severity: .warning, message: message.description, data: data)
+    }
+
+    public static func info(_ message: String, data: CustomStringConvertible? = .none) -> Self {
+        Self(severity: .info, message: message, data: data)
+    }
+
+    public static func info(_ message: CustomStringConvertible, data: CustomStringConvertible? = .none) -> Self {
+        Self(severity: .info, message: message.description, data: data)
+    }
+
+    public static func note(_ message: CustomStringConvertible, data: CustomStringConvertible? = .none) -> Self {
+        // 👀 do we need info and note?
+        Self(severity: .note, message: message.description, data: data)
+    }
+
+    public enum Severity: Equatable {
+        case error
+        case warning
+        // 👀 do we need info and note?
+        case info
```

Comment:
It seems redundant to have both.  But maybe just "remark" since that's useful to record (any hiding of remarks should be in the presentation layer, I think).

Reply:
thanks @abertelrud. typically logging systems of this nature have "info", "trace" and "debug" levels, which are filtered out by the collectors by default, but users can choose verbosity level to see more data. this was my plan here too, as follow on PR.

the main question I have is what to do with the existing "note" and "remark" levels. do they have special meaning in IDEs (eg Xcode) that need to integrate with this API, or can we transition to the more standard levels?

Reply:
I guess what "standard" means depends on the subsystem.  In Xcode's case, the modeling was initially due to GCC's "error:", "warning:", and "note:" distinction, and so it's pretty sparse.  But there is a desire to extend it.

LLVM has Ignored, Note, Remark, Warning, Error, Fatal (https://clang.llvm.org/docs/UsersManual.html#diagnostic-mappings).  So those are the kinds we'll get from Clang and Swift and we'll need to map them to something in the internal subsystem, once we start to parse those files (which TSC has some support for and which I have a start at adopting).  So if we have few levels we'll be losing some fidelity there.

It makes sense to have "info", "trace", and "debug" as levels that are optionally filtered out already at collection time.  For others, such as errors, warnings, and compiler remarks, I think we'd want to always collect them but only conditionally show them (or rather, the filtering out should be on the client side).

Ideally I'd like to not have a predefined set of levels at all but to have a system similar to Universal Type Identifiers where there would be a small built-in set of severities that others could be defined as conforming to.  Level comparison checks would then turn into conformance checks, which is maybe more complexity than we want to deal with at this point.

So I'd say that we can reduce this down to error, warning, and some way of representing the compiler's notion of Remark, whether that's Info or something else.